### DISCUSSION TITLE
Make use of CI/CD for framework development

### DISCUSSION THREAD

**Discussion by knoepfel** [Framework-R-D/phlex#523] [Ideas]
**_DUNE US S&C R&D item 068_**

Robust Framework development will require sufficient use of CI/CD technology.  It is not the responsibility of the framework project to provide the CI/CD infrastructure; however, the development team must develop a CI/CD policy.  The CI/CD technology should be able to support that policy, which may include:

- Ensuring compliance of coding guidelines
- Code coverage metrics
- Successful execution of unit and integration tests

One possibility is to incorporate these checks as GitHub actions, which may or may not be constrained based on our free use of GitHub.

**Comment by pcanal** [comment]
Related issue: https://github.com/Framework-R-D/phlex/issues/73

**Comment by knoepfel** [comment]
We should implement a GitHub CI action before we start sharing code.

**Comment by knoepfel** [comment]
Related issues: 
- https://github.com/Framework-R-D/phlex/issues/73
- https://github.com/Framework-R-D/phlex-design/issues/32

**Comment by knoepfel** [comment]
_[Based on meeting held with @knoepfel, @aolivier23, and @greenc-FNAL]_

## Prioritized list

### Needed now

- [x] 0. Container image used by GitHub PR action, which includes GCC 15, ROOT 6.36/02, etc.
   #32

For PRs, we need to verify that the code:

- [ ] 1. Successfully builds and the unit tests successfully execute for:
	- [x] GCC and Clang
	- [ ] Optimized vs. Debug (with enabled assertions)
	- [ ] AlmaLinux 9 (`x86_64_v3`)
- [ ] 2. Conforms to the approved style guide
- [x] 3. All code changes are 100% covered by code-coverage tests

These checks will be implemented as GitHub actions.  The fixes for any failures are the responsibility of the PR submitters.

We also need:

- [ ] 5. Debugging, benchmarking, and profiling tools
- [ ] 6. Documentation generation for source code

### Needed by January 31, 2026

- [x] 7. Incorporate static analysis facilities as part of PR process
	1. Technology options: `clang-tidy`
	2. Includes conformance to C++ core guidelines
- [ ] 8. Use release facilities for:
	1. Source/binary package production
	2. Dependency management and maintenance
- [ ] 9. Automatically generate and upload released and in-development documentation
- [ ] 10. Plan for AI use (code analysis, generation, and review), given experience in first 6 months
- [ ] 11. Determine whether additional debugging, benchmarking, and profiling tools are required.
    Framework-R-D/phlex#73

### Needed by July 31, 2026

- [ ] 13. We need to be able to access a GPU
- [x] 14. Dependabots that can flag security issues
- [ ] 15. Cross-package PR coordination

**Comment by wddgit** [comment]
My 2 cents. We could add a "Needed Last Week" category that includes "builds and unit tests pass for an optimized GCC build on AlmaLinux 9 (x86_64_v3)". Get that working first.

**Reply by knoepfel** [reply]
Thanks, @wddgit.  Let's talk about this at the next Phlex meeting.

**Comment by wlav** [comment]
Do we pick/have specific compiler versions? Either fixed or minimum? For that matter, vanilla Clang or Apple Clang? And for releases, do we expect (or are we expected) to deliver them as docker images? Is DUNE expected to use Phlex releases, or will they build from source for their own distributions?

**Reply by knoepfel** [reply]
Let's talk about this at the next Phlex meeting.
